### ISSUE TITLE
No hardware decoding on OS X 10.11 (Public Beta 5)

### ISSUE THREAD

**Issue by wzxu** [mpv-player/mpv#2240]
For a while I've been using Movist instead since mpv [can't play in fullscreen through HDMI](https://github.com/mpv-player/mpv/issues/2062) on my TV, but just then decided to come back and check out if things changed. Well, they didn't. But I came across another issue.

In the original config I used vda as hardware decoding API, but now it just gets stuck at
```
 (+) Video --vid=1 (*) (h264)
 (+) Audio --aid=1 --alang=eng (*) (aac)
Using hardware decoding.
AO: [coreaudio] 48000Hz stereo 2ch floats
```
So I searched around, and [found that](https://github.com/mpv-player/mpv/issues/2039) vda was deprecated in 10.11 (yet mpv still tries to use it instead of falling back to software decoding. bug?) and there's a new "video toolbox" setting available.

However, when I set `hwdec=videotoolbox` it always falls back to `Using software decoding.` I wondered for a moment if my configuration doesn't support it, but t'm using the latest OS X on the latest MacBook Pro, so that can't be. Just want to put this out here and see if others notice it as well.

Using mpv installed via `brew install --HEAD --with-official-libass mpv`.

**Comment by wm4** [comment]
Use mpv git master.

Nobody cares about releases anymore.

**Comment by pigoz** [comment]
Btw, you need FFmpeg's git master as well if you want to use videotoolbox support.

**Comment by wzxu** [comment]
Ah, thanks. Reinstall ffmpeg with --HEAD and rebuild mpv again did the trick.

What about the issue where mpv still tries to use via and gets stuck instead of falling back to software decoding, though? Can you reproduce that? I figure it might trouble some people when El Capitan releases and they upgrade.

**Comment by pigoz** [comment]
VDA should not be available at all on 10.11

**Comment by wzxu** [comment]
Yes, that's my point. 

Say, some users (I'd say a lot?) were already using vda in their mpv.conf in 10.10, and when they upgrade to 10.11, vda will not be available. When that happens, mpv should either 1) pop up an alert notifying the user that they should use videotoolbox instead of vda now, or 2) fall back to software decoding silently.

But currently mpv does neither, it just tries to use vda still and gets stuck.

**Comment by pigoz** [comment]
Seems to work here if VDA is not compiled in (it should not be on 10.11 because the framework is missing afaik).

```
./configure --disable-vda-hwaccel
./waf build
pigoz@NAVi ~/dev/mpv ±master » build/mpv --hwdec=vda /Users/pigoz/Movies/The\ Sopranos\ S1/The.Sopranos.S01E01.720p.BluRay.DD5.1.x264-ITSat.mkv
Playing: /Users/pigoz/Movies/The Sopranos S1/The.Sopranos.S01E01.720p.BluRay.DD5.1.x264-ITSat.mkv
 (+) Video --vid=1 (*) (h264)
 (+) Audio --aid=1 (*) (ac3)
 (+) Subs  --sid=1 'The.Sopranos.S01E01.720p.BluRay.DD5.1.x264-ITSat.srt' (subrip) (external)
Using software decoding.
[sub] Using subtitle charset: UTF-8-BROKEN
AO: [coreaudio] 48000Hz stereo 2ch float
VO: [opengl-hq] 1280x720 yuv420p
AV: 00:00:01 / 01:00:01 (0%) A-V:  0.000 Cache: 10s+3MB
```

Were you maybe on 10.11 a binary compiled on 10.10?

**Comment by wzxu** [comment]
No. When I first reported this, mpv and all dependencies were freshly installed via Homebrew.

Now that I reinstalled the git version ffmpeg and thus enabled videotoolbox, mpv would opt for videotoolbox even with hwdec=vda, which kinda solved the problem. Still, I think people (at least those building it themselves) will be confused when they first upgrade to 10.11 and suddenly mpv doesn't work.

There should at least be clearer documentation on this IMHO. Perhaps clarify that vda is deprecated and videotoolbox should be used instead in 10.11 (right now it only says 'newer OS X only' for videotoolbox in the manual), and that git version of ffmpeg is required.

**Comment by pigoz** [comment]
I still don't understand how you can enable VDA at compile time on 10.11 since the headers should be missing.

**Comment by wzxu** [comment]
Alright. Just to make sure I wan't crazy, I uninstalled mpv, along with all its dependencies, cleaned up, and did what I did when I installed it before:
```
brew install --without-x11 cairo
brew install --HEAD --with-official-libass mpv
brew linkapps mpv
```
and then, create a new mpv.conf with only this line in it:
```
hwdec=vda
```
and it gets stuck again in the same place.

---

I think I found the culprit! Just discovered something right before I submitted this. If I run mpv with --msg-level=all=v, then in the output I spotted these lines: 
```
[vd]     lavc:h264_vda (h264) - H.264 (VDA acceleration)
[vd] Opening video decoder lavc:h264
[vo/opengl] Loading hwdec driver 'videotoolbox'
Using hardware decoding.
```
So it looks like mpv does switch to videotoolbox instead of forcing vda like I imagined, but since normal install came with the stable ffmpeg, there's no videotoolbox support, and somehow mpv does't fallback to software decoding in this scenario (it does if I specify hwdec=videotoollbox without git ffmpeg), it gets stuck.

In summary, here's what happens currently if no git version ffmpeg is installed:

* hwdec=videotoolbox
videotoolbox -> *NOT AVAILABLE* -> software

* hwdec=vda
vda -> *NOT AVAILABLE* -> videotoolbox -> *NOT AVAILABLE* -> stuck

**Comment by wm4** [comment]
On the video output level it's always called ``videotoolbox`` (because the same code is used for video output).

What gets stuck?

**Comment by wzxu** [comment]
Do you mean setting hwdec to vda or videotoolbox actually does the same thing? That's even more curious since one falls back to software decoding while the other does not. Weird.

Stuck, as in the terminal output stops at
```
 (+) Video --vid=1 (*) (h264)
 (+) Audio --aid=1 --alang=eng (*) (aac)
Using hardware decoding.
AO: [coreaudio] 48000Hz stereo 2ch floats
```
and no window shows up, and nothing is playing.

**Comment by wm4** [comment]
I meant what I said. The video output is the same, just the decoder is different.

>and no window shows up, and nothing is playing.

That is of course not expected. It indicates the decoder never outputs a frame.
